proto/ptypes/interceptor.go

```diff
@@ -0,0 +1,74 @@
+package ptypes
+
+import (
+	context "context"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+func WithIdentityValidationControllerServerInterceptor(volumeName, instanceName string) grpc.ServerOption {
+	return grpc.UnaryInterceptor(identityValidationServerInterceptor(volumeName, instanceName, "controller"))
+}
+
+func WithIdentityValidationReplicaServerInterceptor(volumeName, instanceName string) grpc.ServerOption {
+	return grpc.UnaryInterceptor(identityValidationServerInterceptor(volumeName, instanceName, "replica"))
+}
+
+func identityValidationServerInterceptor(volumeName, instanceName, serverType string) grpc.UnaryServerInterceptor {
+	// Use a closure to remember the correct volumeName and/or instanceName.
+	return func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if ok {
+			incomingVolumeName, ok := md["volume-name"]
+			// Only refuse to serve if both client and server provide validation information.
+			if ok && volumeName != "" {
+				log := logrus.WithFields(logrus.Fields{"method": info.FullMethod,
+					"clientVolumeName": incomingVolumeName[0], "serverVolumeName": volumeName})
+				if incomingVolumeName[0] != volumeName {
+					log.Error("Invalid gRPC metadata")
+					return nil, status.Errorf(codes.FailedPrecondition, "Incorrect volume name; check %s address",
```

Comment:
NIT: We can add `incomingVolumeName[0]` and `volumeName` to the error info. Same below.